source/base/parameter_handler.cc

```diff
@@ -2311,13 +2319,391 @@ ParameterHandler::print_parameters (std::ostream     &out,
     }
 
   // dive recursively into the subsections
-  print_parameters_section (out, style, 0);
+  recursively_print_parameters (std::vector<std::string>(), // start at the top level
+                                style,
+                                0,
+                                out);
 
   return out;
 }
 
 
 
+void
+ParameterHandler::recursively_print_parameters (const std::vector<std::string> &target_subsection_path,
+                                                const OutputStyle               style,
+                                                const unsigned int              indent_level,
+                                                std::ostream                   &out) const
+{
+  AssertThrow (out, ExcIO());
+
+  Assert ((style != XML) && (style != JSON),
+          ExcMessage ("This function should not be necessary for XML or JSON output."));
+
+  const boost::property_tree::ptree &current_section
+    = entries->get_child (collate_path_string(target_subsection_path));
+
+  unsigned int overall_indent_level = indent_level;
+
+  switch (style)
+    {
+    case Text:
+    case ShortText:
+    {
+      // first find out the longest entry name to be able to align the
+      // equal signs to do this loop over all nodes of the current
+      // tree, select the parameter nodes (and discard sub-tree nodes)
+      // and take the maximum of their lengths
+      std::size_t longest_name = 0;
+      for (boost::property_tree::ptree::const_iterator
+           p = current_section.begin();
+           p != current_section.end(); ++p)
```

Comment:
Hm, not trivial apparently. This doesn't seem to compile:
```
for (boost::property_tree::ptree::const_iterator p : current_section)
```
and nor does
```
for (auto p : current_section)
```
I get error messages that read:
```
/home/fac/f/bangerth/p/deal.II/1/dealii/source/base/parameter_handler.cc:2363:60: error: conversion from ‘const std::pair<const std::basic_string<char>, boost::property_tree::basic_ptree<std::basic_string<char>, std::basic_string<char> > >’ to non-scalar type ‘boost::property_tree::basic_ptree<std::basic_string<char>, std::basic_string<char> >::const_iterator’ requested
       for (boost::property_tree::ptree::const_iterator p : current_section)
```
and
```
/home/fac/f/bangerth/p/deal.II/1/dealii/source/base/parameter_handler.cc:2364:33: error: base operand of ‘->’ has non-pointer type ‘std::pair<const std::basic_string<char>, boost::property_tree::basic_ptree<std::basic_string<char>, std::basic_string<char> > >’                                                                                                
         if (is_parameter_node (p->second) == true)                                                                                                                               
                                 ^                                                                                                                                              
```
Given that some non-trivial conversion appears to be going on here, I'll leave the code as is to make the intent as clear as I can.